core-bundle/src/Resources/contao/drivers/DC_Folder.php

```diff
@@ -391,39 +391,42 @@ public function showAll()
 			}
 			else
 			{
-				$arrRoot = array();
-
-				// Respect existing limitations (root IDs)
-				if (\is_array($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root'] ?? null))
+				$isBasePath = static function (string $path, array $basePaths)
 				{
-					while ($objRoot->next())
+					foreach ($basePaths as $basePath)
 					{
-						foreach ($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root'] as $root)
+						if (Path::isBasePath($basePath, $path))
 						{
-							if (strncmp($root . '/', $objRoot->path . '/', \strlen($root) + 1) === 0)
-							{
-								if ($objRoot->type == 'folder' || empty($this->arrValidFileTypes) || \in_array($objRoot->extension, $this->arrValidFileTypes))
-								{
-									$arrFound[] = $objRoot->path;
-								}
-
-								$arrRoot[] = ($objRoot->type == 'folder') ? $objRoot->path : \dirname($objRoot->path);
-								continue 2;
-							}
+							return true;
 						}
 					}
-				}
-				else
+
+					return false;
+				};
+
+				$arrRoot = array();
+				$root = $GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root'] ?? null;
+
+				while ($objRoot->next())
 				{
-					while ($objRoot->next())
+					// Respect given filemounts, e.g. set in the initPicker() method
+					if ((!empty($this->arrFilemounts) || null !== $root) && !$isBasePath($objRoot->path, $this->arrFilemounts))
```

Comment:
```suggestion
					if (!empty($this->arrFilemounts) && !$isBasePath($objRoot->path, $this->arrFilemounts))
```

The `|| null !== $root` part does not make sense to me. If `$this->arrFilemounts` is an empty array, the second half of the if condition will always be `false`, so `null !== $root` never makes a difference.

`$root` is then handled in the next if condition (line 419), which would also be the case without `null !== $root`.

Reply:
Without `null !== $root` the search would show results outside of the filemounts of the user. Can be reproduced by setting `path` to `files/foo` and filemounts to `[files/bar]` and search for something like `jpg`.